### PR TITLE
docs: WIP add naught coin exercise

### PR DESCRIPTION
**Pull request by ChmielewskiKamil**
This is a draft PR for #207

tldr; 
Where should this exercise be placed? Which number should I give to it?

- [x] Add basic exercise readme
- [ ] Add exercise files
- [ ] Expand the `setup` section
- [ ] Add a number to the exercise name 
- [ ] Reorder previous exercises
- [ ] Proofread once again
- [ ] Correct grammar

It came up to be a little bit wordy compared to other exercises. I wanted to narrow the gap between the first four exercises, which are relatively easy, to the last 4 harder exercises. 

Instead of hints, I decided to go with hidden-by-default steps that the user can follow if he gets stuck. It does not spoil the solution. 

I am publishing this as a draft so you can review it early and tell me if this is the right way to go about it.

Thanks!

### CODE REVIEW

File: program-analysis/echidna/Exercise-X.md [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+    ├── ERC20.sol
+    ├── IERC20.sol
+    └── IERC20Metadata.sol
+```
+
+Depending on your framework, the `NaughtCoin` contract will be in the `src` folder (Foundry) or `contracts` (Hardhat). OpenZeppelin dependencies are in the `level-utils` folder, and all Echidna-related files are in the `crytic` folder.
+
+### Exercise goals
+
+The main goal is to make the player's balance equal to `0`.  You might know the answer right away. However, the purpose of this exercise is to define the whole system using properties in an iterative way. Start small and see what you can find. Expand over time. Here are some steps that you may take:
```
Comment (ChmielewskiKamil) [inline program-analysis/echidna/Exercise-X.md:36]:
```suggestion
The main goal is to make the player's balance equal to `0`.  You might know the answer right away. However, the purpose of this exercise is to describe the whole system using properties in an iterative way. Start small and see what you can find. Expand over time. Here are some steps that you may take:
```

File: program-analysis/echidna/Exercise-X.md [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+<details>
+  <summary>Example state setup</summary>
+
+   ```
+   // SPDX-License-Identifier: MIT
+  pragma solidity ^0.8.0;
+  
+  import {NaughtCoin} from "src/NaughtCoin.sol";
+  
+  contract ExternalTestSimple {
```
Comment (ChmielewskiKamil) [inline program-analysis/echidna/Exercise-X.md:105]:
```suggestion
  contract ExternalTest {
```

As there is no advanced setup followup, remove the word simple.

File: program-analysis/echidna/Exercise-X.md [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+    }
+   ```
+
+</details>
+
+<details>
+  <summary>2nd property</summary>
+
+   ```
+   function sender_balance_is_equal_to_initial_supply() public {
```
Comment (ChmielewskiKamil) [inline program-analysis/echidna/Exercise-X.md:146]:
```suggestion
   function player_balance_is_equal_to_initial_supply() public {
```

Change sender to player

File: program-analysis/echidna/Exercise-X.md [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+  
+  import {NaughtCoin} from "src/NaughtCoin.sol";
+  
+  contract ExternalTestSimple {
+   address player;
+      NaughtCoin public naughtCoin;
+  
+      constructor() {
+       player = msg.sender;
+          naughtCoin = new NaughtCoin(player);
```
Comment (ChmielewskiKamil) [inline program-analysis/echidna/Exercise-X.md:111]:
```suggestion
      naughtCoin = new NaughtCoin(player);
```

Remove spacing

File: program-analysis/echidna/Exercise-X.md [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+
+</details>
+
+<details>
+  <summary>2nd property improved with pre-conditions</summary>
+  
+  In plain English we have defined this invariant as The player token balance should equal the initial supply if the current `block.timestamp < timelock`. The second part of this sentence specifies exactly when this property should be tested.
+  
+   ```
+   function sender_balance_is_equal_to_initial_supply() public {
```
Comment (ChmielewskiKamil) [inline program-analysis/echidna/Exercise-X.md:159]:
```suggestion
   function player_balance_is_equal_to_initial_supply() public {
```

Change sender to player

File: program-analysis/echidna/Exercise-X.md [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+  This property would be invalidated if any of our previous property were invalidated (which seems to not be the case at the moment). Let's leave this property for now and move on.
+
+</details>
+
+<details>
+  <summary>6th property</summary>
+  
+  The token transfer via `transferFrom` should fail if the current `block.timestamp < timelock` and/or spender has enough allowance.
+  
+  This property in its base form is the same as the one with `transfer` function.
```
Comment (ChmielewskiKamil) [inline program-analysis/echidna/Exercise-X.md:302]:
```suggestion
  This property in its base form is the same as the one with the `transfer` function.
```

Add article "the"

File: program-analysis/echidna/Exercise-X.md [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+            );
+        }
+    }
+   ```
+
+Run Echidna and see what you get. Echidna emits the `AssertionFailed` event with the `amount` of `0`. It means that the `transferFrom` succeeded. If you look at the corpus, you will see that our post-conditions were invalidated. Echidna found a way to change player's balance.
+
+The property `playerBalance == InitialSupply` still holds. You need to guide Echidna a little bit more for it to invalidate this property.
+
+Try to get into the mindset of exploring the system you are auditing. Since, we have discovered something new about the system, a `transferFrom` function does not fail, as we initially expected, the next logical step would be to test this functionality and make sure that it works properly.
```
Comment (ChmielewskiKamil) [inline program-analysis/echidna/Exercise-X.md:331]:
```suggestion
Try to get into the mindset of exploring the system you are auditing. Since we have discovered something new about the system, a `transferFrom` function does not fail, as we initially expected, the next logical step would be to test this functionality and ensure that it works properly.
```

Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:331]: Change wording

File: program-analysis/echidna/Exercise-X.md [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+  │   Block delay: 42595                                                                                                 │
+  │ Event sequence:                                                                                                      │
+  │ Panic(1)
+ ```
+
+</details>
+
+#### Fixing a bug
+
+Last step is to fix a bug that you have found. Try to fix it and re-run your properties. Check if they hold.
```
Comment (ChmielewskiKamil) [inline program-analysis/echidna/Exercise-X.md:383]:
```suggestion
The last step is to fix a bug that you have found. Try to fix it and re-run your properties. Check if they hold.
```

Add article "the"

File: program-analysis/echidna/Exercise-X.md [hunk 9]
```diff
@@ -0,0 +1,440 @@
...
+Add a `Player(address player)` event to your contract and emit it in your properties. Look at the call sequence.
+
+</details>
+
+<details>
+  <summary>Fixing a bug part 2</summary>
+  
+The player's address is `0x0000...3000`. Echidna is making calls from multiple accounts. It was able to increase the allowance of an address `0x000...1000` and make a call to `transferFrom(player, 0x000...1000, amount)`. This is expected!
+
+The `lockTokens` modifier does not prevent others from making transfers, only the player is constrained.
```
Comment (ChmielewskiKamil) [inline program-analysis/echidna/Exercise-X.md:409]:
```suggestion
The `lockTokens` modifier does not prevent others from making transfers. Only the player is constrained.
```

Split sentences

File: program-analysis/echidna/Exercise-X.md [hunk 10]
```diff
@@ -0,0 +1,440 @@
...
+If you run Echidna, you will see that the properties still fail. Why is that?
+
+Add a `Player(address player)` event to your contract and emit it in your properties. Look at the call sequence.
+
+</details>
+
+<details>
+  <summary>Fixing a bug part 2</summary>
+  
+The player's address is `0x0000...3000`. Echidna is making calls from multiple accounts. It was able to increase the allowance of an address `0x000...1000` and make a call to `transferFrom(player, 0x000...1000, amount)`. This is expected!
```
Comment (ChmielewskiKamil) [inline program-analysis/echidna/Exercise-X.md:407]:
```suggestion
The player's address is `0x0000...3000`. Echidna is making calls from multiple accounts. It was able to increase the allowance of an address `0x000...1000` and make a call to `transferFrom(player, 0x000...1000, amount)`. This is expected!

[This happens because of the `multi-abi`](https://github.com/crytic/building-secure-contracts/blob/master/program-analysis/echidna/using-multi-abi.md). 
```

File: program-analysis/echidna/Exercise-X.md [hunk 11]
```diff
@@ -0,0 +1,442 @@
...
+ 43 |     |         // post-conditions
+ 44 |     |         assert(
+ 45 |     |             naughtCoin.balanceOf(player) == playerBalanceBefore &&
+ 46 |     |                 naughtCoin.balanceOf(bob) == bobBalanceBefore
+ 47 |     |         );
+ 48 |     |     }
+ 49 |     | }
+   ```
+
+   You can see that the transfer reverted as expected. But... Our post-conditions weren't checked. If you have read the [ERC20 spec](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-20.md), you will know that the `transfer` function returns a status `boolean`. In the case of success, we can emit a special `AssertionFailed` event that won't stop the execution of the function. This way, our post-conditions will be checked.
```
Comment (glarregay-tob) [inline program-analysis/echidna/Exercise-X.md:228]: What is your opinion on using `try`/`catch` instead of relying on the return value? OZ tokens either revert or return `true`:
 ```
if (currentTime < naughtCoin.timeLock()) {
         // actions
         try naughtCoin.transfer(bob,amount) { 
              // call succeeded, this is not ok
              assert(false);
         } catch {
              // call reverted
         }
}
```

File: program-analysis/echidna/Exercise-X.md [hunk 12]
```diff
@@ -0,0 +1,442 @@
...
+   ```
+
+Run Echidna and check the corpus. Now we have fully covered this property.
+
+</details>
+
+<details>
+  <summary>4th property</summary>
+  
+  According to the ERC20 spec, the `approve` function should not fail if the caller has enough tokens to make the approval.
```
Comment (glarregay-tob) [inline program-analysis/echidna/Exercise-X.md:266]: `approve` won't fail even if you don't have enough tokens. It's like signing a blank check: "I allow this account to spend up to xxx of my funds in one or more transactions"
It might happen that I don't have the funds yet, but the approval goes through anyway.

File: program-analysis/echidna/Exercise-X.md [hunk 13]
```diff
@@ -0,0 +1,442 @@
...
+
+</details>
+
+<details>
+  <summary>5th property</summary>
+  
+  Player should not `burn` tokens before the `timeLock` period.
+  What is token burning?
+  
+  You can burn tokens by sending them to the `0` address or by sending them to a non existent address. Sending to the `0` address is not possible because of the `0` address checks in the `ERC20` OZ standard. Sending them to any address shouldn't be possible as well because of the `timeLock`.
```
Comment (glarregay-tob) [inline program-analysis/echidna/Exercise-X.md:291]: This is a bit of a technicality, but transferring it to anywhere doesn't burn them, it just locks them.
A "proper" burning should decrease the total supply of the token, and not all tokens implement public functions to do so. Base [ERC20 OZ token](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/a70ee4e3bbee565e49217c9ec59de5354f8af9d8/contracts/token/ERC20/ERC20.sol#L277-L288) implements an internal _burn method that can be used by token implementations.

File: program-analysis/echidna/Exercise-X.md [hunk 14]
```diff
@@ -0,0 +1,442 @@
...
+  This property would be invalidated if any of our previous property were invalidated (which seems to not be the case at the moment). Let's leave this property for now and move on.
+
+</details>
+
+<details>
+  <summary>6th property</summary>
+  
+  The token transfer via `transferFrom` should fail if the current `block.timestamp < timelock` and/or spender has enough allowance.
+  
+  This property in its base form is the same as the one with the `transfer` function.
```
Comment (glarregay-tob) [inline program-analysis/echidna/Exercise-X.md:302]: If the previous one is changed to `try`/`catch`, this one should be changed too

File: program-analysis/echidna/Exercise-X.md [hunk 15]
```diff
@@ -0,0 +1,442 @@
...
+If you run Echidna, you will see that the properties still fail. Why is that?
+
+Add a `Player(address player)` event to your contract and emit it in your properties. Look at the call sequence.
+
+</details>
+
+<details>
+  <summary>Fixing a bug part 2</summary>
+  
+The player's address is `0x0000...3000`. Echidna is making calls from multiple accounts. It was able to increase the allowance of an address `0x000...1000` and make a call to `transferFrom(player, 0x000...1000, amount)`. This is expected!
```
Comment (glarregay-tob) [inline program-analysis/echidna/Exercise-X.md:407]: Visual tip: We usually write `0x30000` or `0x10000` to avoid the trailing zeros.

File: program-analysis/echidna/Exercise-X.md [hunk 16]
```diff
@@ -0,0 +1,442 @@
...
+### Exercise setup
+
+The `NaughtCoin` contract has several dependencies. The directory structure that you should aim for looks like this:
+
+```
+src
+├── NaughtCoin.sol
+├── crytic
+│   ├── ExternalTest.sol
+│   └── config.yaml
```
Comment (glarregay-tob) [inline program-analysis/echidna/Exercise-X.md:24]: Can you expand a bit on what the config should look like?

File: program-analysis/echidna/Exercise-X.md (hunk 11 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:228]: Thank you so much for your feedback, @glarregay-tob! As I described in my [blog post](https://wizzardhat.com/ethernaut-level-15-naughtcoin/), I initially went for the `try`/`catch`. They are much more intuitive and overall way better than return value checks. I switched to value checks for a couple of reasons:
- I couldn't get Echidna to work with `try`/`catch`. The property was either always passing or always reverting
- That's what Justin did in the [Intro to AMM's invariants workshop](https://www.youtube.com/watch?v=n0RaKKVTGvA), and I thought that it might be a better practice
- I must admit that in this particular case, the `try` / `catch` with an empty catch block, as explained in the [How to write good properties](https://github.com/crytic/building-secure-contracts/blob/master/program-analysis/echidna/property-creation.md) section is a little bit confusing to me. 

I assumed that the OZ standard returns boolean values, but as you pointed out, this is wrong. The `transfer` and `transferFrom` functions return a boolean value, but the internal `_transfer` function reverts on failure. 

Thanks again for pointing that out; I will also fix this in my article. 

I will re-write the properties with `try`/`catch` and update the exercise.

File: program-analysis/echidna/Exercise-X.md (hunk 12 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:266]: This is an oversight on my part. It's called an "allowance overrun", right? 

Will fix that. Thanks!

File: program-analysis/echidna/Exercise-X.md (hunk 13 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:291]: This is a fair point. I think the word "effectively" is missing. Sending tokens to a non-existing address locks them up, which effectively burns them. 

The `ERC20` used in this Ethernaut challenge did not implement a public burn functionality.

File: program-analysis/echidna/Exercise-X.md (hunk 15 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:407]: Great tip, will fix it. Thanks!

File: program-analysis/echidna/Exercise-X.md (hunk 16 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:24]: I will expand on it definitely, along with the expanded setup configuration.

File: program-analysis/echidna/Exercise-X.md (hunk 11 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:228]: I think it might be helpful to leave this section as it is and create a new follow-up section with Justin's setup using an external `Setup` contract and re-write properties using `try`/`catch` to showcase that both options are valid. What do you think, @glarregay-tob?

File: program-analysis/echidna/Exercise-X.md (hunk 12 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:266]: Does it mean that this property is not the best one to test? If technically it will never fail?

File: program-analysis/echidna/exercises/exerciseX/NaughtCoin.sol [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    // string public constant name = 'NaughtCoin';
+    // string public constant symbol = '0x0' -> 'NTC';
+    // uint public constant decimals = 18;
+    uint256 public timeLock = block.timestamp + 10 * 365 days;
+    uint256 public INITIAL_SUPPLY;
+    address public player;
+
+    // Note that the original symbol is changed from 0x0 to NTC
+    // This is caused by https://github.com/crytic/echidna/issues/909
+    constructor(address _player) ERC20("NaughtCoin", "NTC") {
```
Comment (ChmielewskiKamil) [inline program-analysis/echidna/exercises/exerciseX/NaughtCoin.sol:16]: The `ERC20` symbol had to be changed because of https://github.com/crytic/echidna/issues/909

File: program-analysis/echidna/Exercise-X.md (hunk 16 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:24]: Section added

File: program-analysis/echidna/Exercise-X.md (hunk 12 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:277]: Fixed

File: program-analysis/echidna/Exercise-X.md (hunk 13 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md]: Fixed

File: program-analysis/echidna/Exercise-X.md (hunk 15 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md]: Fixed

File: program-analysis/echidna/Exercise-X.md (hunk 11 above)
Comment (glarregay-tob) [reply program-analysis/echidna/Exercise-X.md:239]: It is weird that it didn't work for you. Let's go with the option that works first, and then if you want you can add the new section.
As a side note, this can actually be generic for reverting and non reverting tokens: you can get the return value from a `try` call by using `try contract.function(param) returns (bool success) {` if that helps. 
Maybe "How to write good properties" needs an update too :) thanks for pointing it out!

File: program-analysis/echidna/Exercise-X.md (hunk 12 above)
Comment (glarregay-tob) [reply program-analysis/echidna/Exercise-X.md:277]: I'm not aware of it having a special name, the ERC20 standard does not mention it. You just allow the account to spend your funds.
Reviewing OZ's implementation, the only way it can fail is if the address zero is involved, otherwise it just updates the allowance.

File: program-analysis/echidna/Exercise-X.md (hunk 11 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:239]: > As a side note, this can actually be generic for reverting and non reverting tokens: you can get the return value from a try call by using try contract.function(param) returns (bool success) { if that helps. 

This might actually be the case. The `transfer` and `transferFrom` functions return true if they succeed. The reverting mechanism is caused by the internal function `_transfer`. 

I will test it out in the follow-up section. 

Thanks @glarregay-tob!

File: program-analysis/echidna/Exercise-X.md (hunk 12 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md]: Fixed

File: program-analysis/echidna/Exercise-X.md (hunk 16 above)
Comment (glarregay-tob) [reply program-analysis/echidna/Exercise-X.md:24]: This is great now. Keep in mind that you can add some links to `using-multi-abi.md` and `testing-modes.md` to avoid some explanations.

Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:24]: Forgot to do so. I am on that!

**Comment by ChmielewskiKamil** [comment]
As a side note @glarregay-tob. Where should this exercise be placed?

File: program-analysis/echidna/Exercise-X.md (hunk 16 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:24]: Links added.

**Comment by glarregay-tob** [comment]
I think @ggrieco-tob will have a more accurate answer, but here's my take. Having exercise numbers can make it easier to organize them and give people a feeling of progression when someone is following the tutorials/examples. But since the increasing numbers aren't associated with an increased difficulty or anything anymore, we might need to review this approach.

However, for this particular case, I'm a bit on the line if we should consider this a new exercise (because the explanation and the way it is presented is definitely a tutorial, it makes a lot of sense) or if we should open a new repo for Ethernaut solutions using Echidna.

**Comment by ChmielewskiKamil** [comment]
> However, for this particular case, I'm a bit on the line if we should consider this a new exercise (because the explanation and the way it is presented is definitely a tutorial, it makes a lot of sense) or if we should open a new repo for Ethernaut solutions using Echidna.

Yes, it definitely came out to be more of a tutorial than an exercise. I think that splitting the content into [tutorials, how-to guides, exercises, explanations and references](https://documentation.divio.com/introduction/) would be a good long-term approach.

File: program-analysis/echidna/Exercise-X.md (hunk 11 above)
Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:241]: I've been experimenting with the `try`/`catch`, and this is where I got stuck the last time. 

According to [writing good properties step by step](https://github.com/crytic/building-secure-contracts/blob/master/program-analysis/echidna/property-creation.md#enhacing-postcondition-checks), if we expect a function to revert, we should place the `assert(false)` in the first block. It means that the function call should not succeed, and when it does, we are bringing that up with the `assert(false)`. The following code results in a property violation. 

In this property, we expect that the `transfer` will fail.
```solidity
function token_transfer_should_fail_before_timelock_period(uint256 amount)
        public
    {
        amount = _between(amount, 0, token.INITIAL_SUPPLY());

        // pre-conditions
        uint256 currentTime = block.timestamp;
        uint256 playerBalanceBefore = token.balanceOf(address(player));
        uint256 bobBalanceBefore = token.balanceOf(address(bob));

        if (currentTime < token.timeLock()) {
            // action
            try
                player.proxy(
                    address(token),
                    abi.encodeWithSignature(
                        "transfer(address,uint256)",
                        address(bob),
                        amount
                    )
                )
            {
                assert(false);
            } catch {
                /* reverted */
            }

            // post-conditions
            assert(token.balanceOf(address(player)) == playerBalanceBefore);
            assert(token.balanceOf(address(bob)) == bobBalanceBefore);
        }
    }
```

And this Echidna's event sequence:
```
Call sequence:                                                                                                    
1.token_transfer_should_fail_before_timelock_period(0)                                                            
Event sequence:                                                                                                      
Panic(1)                                                                                                            
merror Revert 0x  
```

The `assert(false)` gets reverted. 

What does the `merror Revert 0x` mean?

I got this property working by applying your suggestion about adding the `returns(bool success)`:

```solidity
try
  player.proxy(
      address(token),
      abi.encodeWithSignature(
          "transfer(address,uint256)",
          address(bob),
          amount
      )
  )
returns (bool success, bytes memory returnData) {
  assert(!success);
} catch {
  /* reverted */
}
```

This, however, leaves me with an unused second return parameter, `returnData`. Should I ignore this warning, or is there a workaround? 

The compiler won't let me ignore the second returned data type. The `returns(bool success)` without the `bytes memory returnData` results in: `TypeError: Function returns 2 values, but returns clause has 1 variables.` 

Emitting this event also does not make sense as it won't be seen by Echidna in any case. 
```solidity
event Log(bytes data);

// ...
returns (bool success, bytes memory returnData) {
    emit Log(returnData);
    assert(!success);
} catch {
    /* reverted */
}
```

I might be missing something obvious.

File: program-analysis/echidna/exercises/exerciseX/NaughtCoin.sol (hunk 17 above)
Comment (ggrieco-tob) [reply program-analysis/echidna/exercises/exerciseX/NaughtCoin.sol:16]: Oh, this is fine, thanks for documenting this.

File: program-analysis/echidna/Exercise-X.md (hunk 11 above)
Comment (ggrieco-tob) [reply program-analysis/echidna/Exercise-X.md:241]: > What does the merror Revert 0x mean?

I believe this is caused by calling an external contract with the 0x0 address. Are you sure all the addresses are initialized?

Comment (ChmielewskiKamil) [reply program-analysis/echidna/Exercise-X.md:241]: Hi @ggrieco-tob, Thanks for your response! I took a week off for a winter break. I will test this out as soon as I get back. Thanks 🎉

**Comment by ChmielewskiKamil** [comment]
@glarregay-tob, @ggrieco-tob, I've added the follow-up section with the external `Setup.sol` contract. 

The only thing left is to change the name of this tutorial/exercise and place it in an appropriate place. 

I can reorder other exercises and place this one as `exercise 5`. 
Another way to handle this would be to place it under the `Tutorials` section and call it something like `Writing properties for ERC20`

**Comment by ggrieco-tob** [comment]
@montyly what do you think about the name and reordering of these exercises?
